File: src/net/dinkla/raytracer/math/Vector3D.scala
package net.dinkla.raytracer.math

/**
 * Created by Dinkla on 10.07.2015.
 */
abstract class Vector3D[@specialized(Float, Double) F](x: F, y: F, z: F)(implicit numType: Fractional[F]) extends Element3D[F](x, y, z) {

  def unary_-(): Vector3D[F]

  // unary minus
  def negate(): Vector3D[F] = -this

  def +(v: Vector3D[F]): Vector3D[F] = plus(v)

  // Vector addition gives Vector
  def plus(v: Vector3D[F]): Vector3D[F] // Vector addition gives Vector

  def -(v: Vector3D[F]): Vector3D[F] = plus(-v)

  // Vector addition gives Vector
  def minus(v: Vector3D[F]): Vector3D[F] = plus(-v) // Vector addition gives Vector

  def *(f: F): Vector3D[F] = mult(f)

  def mult(f: F): Vector3D[F]

  def cross(v: Vector3D[F]): Vector3D[F] // cross product

  def dot(v: Vector3D[F]): F // dot product

  def toPoint: Point3D[F]

  def normalize: Vector3D[F] // the normal vector with length 1

  override def equals(obj: scala.Any): Boolean
    = obj match {
      case that: Vector3D[F] => x == that.x && y == that.y && z == that.z
      case _ => false
    }

  override def hashCode(): Int = (x, y, z).hashCode()

}

object Vector3D {

  def Vector3D(x: Double, y: Double, z: Double) = new Vector3DD(x, y, z)

  def Vector3D(x: Float, y: Float, z: Float) = new Vector3DF(x, y, z)

  val ZERO: Vector3DF = new Vector3DF(0, 0, 0)
  val UP: Vector3DF = new Vector3DF(0, 1, 0)
  val DOWN: Vector3DF = new Vector3DF(0, -1, 0)
  val JITTER: Vector3DF = new Vector3DF(0.0072f, 1.0f, 0.0034f)
  val BACK: Vector3DF = new Vector3DF(0, 0, -1)

}

File: src/jvmMain/kotlin/net/dinkla/raytracer/math/Vector3D.kt
package net.dinkla.raytracer.math

import kotlin.math.sqrt

data class Vector3D(val x: Double, val y: Double, val z: Double) {

    constructor(p: Point3D) : this (p.x, p.y, p.z)

    operator fun plus(v: Vector3D): Vector3D = Vector3D(x + v.x, y + v.y, z + v.z)

    operator fun minus(v: Vector3D): Vector3D = Vector3D(x - v.x, y - v.y, z - v.z)

    operator fun times(s: Double): Vector3D = Vector3D(s * x, s * y, s * z)

    infix fun dot(v: Vector3D): Double = x * v.x + y * v.y + z * v.z

    infix fun dot(v: Normal?): Double {
        return if (null == v) x * y * z
        else x * v.x + y * v.y + z * v.z
    }

    infix fun cross(v: Vector3D): Vector3D = Vector3D(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x)

    fun normalize(): Vector3D {
        val l = length()
        return Vector3D(x / l, y / l, z / l)
    }

    operator fun unaryMinus() = Vector3D(-x, -y, -z)

    fun volume(): Double = x * y * z

    fun sqrLength(): Double = x * x + y * y + z * z

    fun length(): Double = sqrt(sqrLength())

    override fun toString(): String = "($x,$y,$z)"

    companion object {
        val ZERO = Vector3D(0.0, 0.0, 0.0)
        val RIGHT = Vector3D(1.0, 0.0, 0.0)
        val UP = Vector3D(0.0, 1.0, 0.0)
        val DOWN = Vector3D(0.0, -1.0, 0.0)
        val FORWARD = Vector3D(0.0, 0.0, 1.0)
        val BACK = Vector3D(0.0, 0.0, -1.0)
        val JITTER = Vector3D(0.0072, 1.0, 0.0034)
    }

}

operator fun Double.times(v: Vector3D) = Vector3D(this * v.x, this * v.y, this * v.z)
